src/btree/bt_sync.c

```diff
@@ -297,8 +209,133 @@ __sync_delete_obsolete_ref(WT_SESSION_IMPL *session, WT_REF *ref)
+    if (previous_state == WT_REF_DELETED) {
+        page_del = ref->page_del;
+        if (page_del == NULL ||
+          __wt_txn_visible_all(session, page_del->txnid, page_del->durable_timestamp)) {
+            WT_RET(__wt_page_parent_modify_set(session, ref, true));
+            __wt_verbose(session, WT_VERB_CHECKPOINT_CLEANUP,
+              "%p: marking obsolete deleted page parent dirty", (void *)ref);
+        } else
+            __wt_verbose(
+              session, WT_VERB_CHECKPOINT_CLEANUP, "%p: skipping deleted page", (void *)ref);
+        return (0);
+    }
+
+    /*
+     * If the page is on-disk and obsolete, mark the page as deleted and also set the parent page as
+     * dirty. This is to ensure the parent is written during the checkpoint and the child page
+     * discarded.
+     */
+    newest_stop_durable_ts = newest_stop_ts = WT_TS_NONE;
+    newest_stop_txn = WT_TXN_NONE;
+    obsolete = false;
+
+    /*
+     * There should be an address, but simply skip any page where we don't find one. Also skip the
+     * pages that have overflow keys as part of fast delete flow. These overflow keys pages are
+     * handled as an in-memory obsolete page flow.
+     */
+    if (__wt_ref_addr_copy(session, ref, &addr) && addr.type == WT_ADDR_LEAF_NO) {
+        /*
+         * Max stop timestamp is possible only when the prepared update is written to the data
+         * store.
+         */
+        newest_stop_ts = addr.ta.newest_stop_ts;
+        newest_stop_durable_ts =
+          addr.ta.newest_stop_ts == WT_TS_MAX ? WT_TS_MAX : addr.ta.newest_stop_durable_ts;
+        newest_stop_txn = addr.ta.newest_stop_txn;
+        obsolete = __wt_txn_visible_all(session, newest_stop_txn, newest_stop_durable_ts);
+    }
+
+    if (obsolete) {
+        WT_RET(__wt_page_parent_modify_set(session, ref, true));
+        WT_STAT_CONN_DATA_INCR(session, cc_pages_removed);
+    }
+
+    __wt_verbose_debug2(session, WT_VERB_CHECKPOINT_CLEANUP,
+      "%p on-disk page obsolete check: %s"
+      "obsolete, stop time point %s",
+      (void *)ref, obsolete ? "" : "not ",
+      __wt_time_point_to_string(
+        newest_stop_ts, newest_stop_durable_ts, newest_stop_txn, tp_string));
+    return (0);
+}
+
+/*
+ * __sync_delete_obsolete_ref --
+ *     Check whether the ref is obsolete according to the newest stop time point and handle the
+ *     obsolete page by either remove it or mark it for urgent eviction.
+ */
+static int
+__sync_delete_obsolete_ref(WT_SESSION_IMPL *session, WT_REF *ref)
+{
+    WT_DECL_RET;
+    uint8_t previous_state;
+    bool busy;
+
+    /* Ignore root pages as they can never be deleted. */
+    if (__wt_ref_is_root(ref)) {
+        __wt_verbose_debug2(
+          session, WT_VERB_CHECKPOINT_CLEANUP, "%p: skipping root page", (void *)ref);
+        return (0);
+    }
+
+    /* Ignore internal pages, these are taken care of during reconciliation. */
+    if (F_ISSET(ref, WT_REF_FLAG_INTERNAL)) {
```

Comment:
How to delete internal pages that have all the children obsolete?

Reply:
While reconciling the internal page if all the leaf pages are obsolete, we do not write this internal page to disk due to no leaf pages, this internal page is also ignored while writing it is a parent internal page. This way the internal page also gets freed, but it is handled via normal reconciliation flow.

Reply:
Thanks